src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs

```diff
@@ -415,9 +437,54 @@ private async Task<bool> TryReadPrefaceAsync()
             {
                 if (!readableBuffer.IsEmpty)
                 {
-                    if (ParsePreface(readableBuffer, out consumed, out examined))
+                    switch (state)
                     {
-                        return true;
+                        case ReadPrefaceState.Preface:
+                            if (readableBuffer.Length >= ClientPreface.Length)
+                            {
+                                if (IsPreface(readableBuffer, out consumed, out examined))
+                                {
+                                    return true;
+                                }
+                                else
+                                {
+                                    // With TLS, ALPN should have already errored if the wrong HTTP version is used.
+                                    // Only perform additional validation if endpoint doesn't use TLS.
+                                    var tlsFeature = ConnectionFeatures.Get<ITlsHandshakeFeature>();
+                                    if (tlsFeature == null)
+                                    {
+                                        state = ReadPrefaceState.Http1x;
+                                        goto case ReadPrefaceState.Http1x;
+                                    }
+
+                                    throw new Http2ConnectionErrorException(CoreStrings.Http2ErrorInvalidPreface, Http2ErrorCode.PROTOCOL_ERROR);
+                                }
+                            }
+                            break;
+                        case ReadPrefaceState.Http1x:
+                            if (ParseHttp1x(readableBuffer, out var detectedVersion))
+                            {
+                                if (detectedVersion == HttpVersion.Http10 || detectedVersion == HttpVersion.Http11)
+                                {
+                                    Log.PossibleInvalidHttpVersionDetected(ConnectionId, HttpVersion.Http2, detectedVersion);
+
+                                    // TODO: Cache bytes
+                                    await _context.Transport.Output.WriteAsync(Encoding.UTF8.GetBytes(@"HTTP/1.1 400 Bad Request
+Connection: close
+
```

Comment:
```suggestion
Connection: close
Content-Type: text/plain
Content-Length: 57

An HTTP/1.x request was sent to an HTTP/2 only endpoint.
```

Reply:
What about server and date headers? Is it ok to leave them out? They're both dynamic: the date changes and server header can be disabled with configuration.

Reply:
Server is optional, let's exclude it for simplicity. Date is technically required, but with many exclusions for error cases, so I wouldn't worry about it right now.

https://datatracker.ietf.org/doc/html/rfc7231#section-7.1.1.2